# Note:
- Since on multiple train and test sets of data we get multiple accuracies by applying one ML algorithm.
- Now it possible that i can get even better accuracy when i check for the same multiple train and test set with different set of algorithm.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [2]:
data = pd.read_csv("Health_Insurance.csv")
data.head()

Age  Purchased
0   19          0
1   35          0
2   26          0
3   27          0
4   19          0

In [3]:
data.shape

(400, 2)

# Seperate X and y

In [4]:
X = data.drop(columns = "Purchased")
y = data['Purchased']

# Split the data into train and test set

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

# Apply Logistic Regression on train set

In [6]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr

LogisticRegression()

In [7]:
lr.fit(X_train, y_train)

LogisticRegression()

# Perform predictions on X-test

In [9]:
y_pred_lr = lr.predict(X_test)
y_pred_lr

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0], dtype=int64)

# Check Accuracy

In [10]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred_lr)

0.9

# Apply KNN Classifier on the train test

In [11]:
from sklearn.neighbors import KNeighborsClassifier
knc = KNeighborsClassifier()
knc

KNeighborsClassifier()

In [12]:
knc.fit(X_train, y_train)

KNeighborsClassifier()

# Perform predictions using KNeighborsClassifier

In [13]:
y_pred_knc = knc.predict(X_test)
y_pred_knc

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0], dtype=int64)

# Check accuracy

In [14]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred_knc)

0.85

# Observation:
- We clearly see that Logistic Regression is giving better accuracy as compared to K Neighbors Classifier

# Check using Logistic Regression whether 90% accuracy is the highest accuracy that this data can have

In [15]:
from sklearn.model_selection import cross_val_score
cross_val_score(lr, X, y, cv = 10)

array([0.95 , 0.65 , 0.825, 0.925, 0.925, 0.825, 0.75 , 0.75 , 0.8  ,
       0.975])

# Observation:
- We observe that using logistic Regression alone, some train test set are giving an accuracy of around 75% and some train and test sets are giving an accuracy of around 97.5% 

In [16]:
from sklearn.model_selection import cross_val_score
cross_val_score(knc, X, y, cv = 10)

array([0.9  , 0.65 , 0.8  , 0.8  , 0.95 , 0.775, 0.75 , 0.725, 0.75 ,
       0.85 ])

# Observations:
- We can see that when k - Neighbors Classifier is applied on the this data with 10 fold cross validation, 95% accuracy is the highest one. Clearly, The accuracy is not crossing 97.5% which was obtained through Logistic Regression